### PR TITLE
Change the ComInterfaceGenerator to output source in one file per interface and use file-scoped types

### PR DESCRIPTION
**Pull request by jkoritzinsky**
This PR changes our codegen to match the same shape of code as in https://github.com/AaronRobinsonMSFT/ComObjectRedux/pull/7

For each interface with `[GeneratedComInterface]`, one file will be generated with the following:

- An `InterfaceInformation` file-scoped type that defines the implementation of `IIUnknownInterfaceType` for the interface
- An `InterfaceImplementation` file-scoped type, defined in the following 3 `partial` definitions for simplicity (we can combine these if we so wish):
  - A DIM-based implementation of the user-defined interface for "managed->unmanaged" scenarios.
  - The stubs for the "unmanaged->managed" scenarios
  - ,A method that allocates the vtable for the type (called by the `InterfaceInformation` type)
- A partial definition of the user-defined interface with the `IUnknownDerivedAttribute` applied, using the above two file-scoped types.

As part of getting this PR working, the following general improvements to the ComInterfaceGenerator project and the tests were applied:

- The implementation code for generating stubs and the vtable allocation was refactored out of `VtableIndexStubGenerator` and generalized to fit the two (now quite different) shapes of the wrapping code.
- The `NativeInterfaceUsage` portion of our code snippets has now been moved into a specific test in the "shape validation" suite for the `VtableIndexStubGenerator` as the validation it does is no longer valid for the `ComInterfaceGenerator` and providing a separate single test does a better job at validating the intention of the check.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs
Comment (jtschuster) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs:309]: Is this something we'll want to switch to using the SyntaxFactory methods, or should it stay as an interpolated string?

Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs:309]: I think this is a lot cleaner as an interpolated string and making it typed with SyntaxFactory doesn't get us the benefits of syntax nodes that we get in the rest of the project (more structure around wrapping syntax and knowing what type of syntax we're working with at various API boundaries). I'd be curious what @AaronRobinsonMSFT and @elinor-fung think for this case.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SyntaxEquivalentNode.cs
Comment (jtschuster) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SyntaxEquivalentNode.cs:16]: I know the singleton is usually named Instance, but in this case I first thought this would compare the SyntaxNodes by reference/instance equality. I'm fine leaving it, but if you think it might be confusing too, it could be worth renaming.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SyntaxEquivalentNode.cs
Comment (jtschuster) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SyntaxEquivalentNode.cs:19]: SyntaxEquivalentComparer.Instance.GetHashCode(Node) throws UnreachableException. I feel like we should either implement that method or throw here.

**Review by jtschuster — APPROVED** [review]
Looks good, just a couple nits/questions. Thank you!

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SyntaxEquivalentNode.cs
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SyntaxEquivalentNode.cs:16]: We can rename it, but I'd rather wait until after this PR is merged in.

**Pull request merged**
